x-pack/elastic-agent/pkg/composable/providers/kubernetesleaderelection/kubernetes_leaderelection.go

```diff
@@ -0,0 +1,163 @@
+	cancelLeaderElection context.CancelFunc
+}
+
+// DynamicProviderBuilder builds the dynamic provider.
+func DynamicProviderBuilder(logger *logger.Logger, c *config.Config) (composable.DynamicProvider, error) {
+	var cfg Config
+	if c == nil {
+		c = config.New()
+	}
+	err := c.Unpack(&cfg)
+	if err != nil {
+		return nil, errors.New(err, "failed to unpack configuration")
+	}
+	return &dynamicProvider{logger, &cfg, nil, nil, nil}, nil
+}
+
+// Run runs the leaderelection dynamic provider.
+func (p *dynamicProvider) Run(comm composable.DynamicProviderComm) error {
+	client, err := kubernetes.GetKubernetesClient(p.config.KubeConfig)
+	if err != nil {
+		// info only; return nil (do nothing)
+		p.logger.Debugf("Kubernetes leaderelection provider skipped, unable to connect: %s", err)
+		return nil
+	}
+	if p.config.LeaderLease == "" {
+		p.logger.Debugf("Kubernetes leaderelection provider skipped, unable to define leader lease")
+		return nil
+	}
+
+	agentInfo, err := info.NewAgentInfo()
+	if err != nil {
+		return err
+	}
+	var id string
+	podName, found := os.LookupEnv("POD_NAME")
+	if found {
+		id = "elastic-agent-leader-" + podName
+	} else {
+		id = "elastic-agent-leader-" + agentInfo.AgentID()
+	}
+
+	ns, err := kubernetes.InClusterNamespace()
+	if err != nil {
+		ns = "default"
+	}
+	lease := metav1.ObjectMeta{
+		Name:      p.config.LeaderLease,
+		Namespace: ns,
+	}
+	metaUID := lease.GetObjectMeta().GetUID()
+	p.leaderElection = &leaderelection.LeaderElectionConfig{
+		Lock: &resourcelock.LeaseLock{
+			LeaseMeta: lease,
+			Client:    client.CoordinationV1(),
+			LockConfig: resourcelock.ResourceLockConfig{
+				Identity: id,
+			},
+		},
+		ReleaseOnCancel: true,
+		LeaseDuration:   15 * time.Second,
+		RenewDeadline:   10 * time.Second,
+		RetryPeriod:     2 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				p.logger.Debugf("leader election lock GAINED, id %v", id)
+				p.startLeading(string(metaUID))
+			},
+			OnStoppedLeading: func() {
+				p.logger.Debugf("leader election lock LOST, id %v", id)
+				p.stopLeading(string(metaUID))
+			},
+		},
+	}
+	ctx, cancel := context.WithCancel(context.TODO())
+	p.cancelLeaderElection = cancel
+	p.comm = comm
+	p.startLeaderElector(ctx)
+
+	return nil
+}
+
+// startLeaderElector starts a Leader Elector in the background with the provided config
+func (p *dynamicProvider) startLeaderElector(ctx context.Context) {
+	le, err := leaderelection.NewLeaderElector(*p.leaderElection)
+	if err != nil {
+		p.logger.Errorf("error while creating Leader Elector: %v", err)
+	}
+	p.logger.Debugf("Starting Leader Elector")
+	go le.Run(ctx)
+}
+
+func (p *dynamicProvider) startLeading(metaUID string) {
+	mapping := map[string]interface{}{
+		"leader": true,
+	}
+
+	processors := []map[string]interface{}{
+		{
+			"add_fields": map[string]interface{}{
+				"fields": mapping,
+				"target": "kubernetes_leaderelection",
+			},
+		},
+	}
```

Comment:
Is this adding the `kubernetes_leaderelection.leader` field to events? I don't think that this info is relevant to users. In any case it would be always set to true, right?
I would suggest to remove this.

Reply:
Yeah I agree, I think that the processors addition should be removed.

Reply:
👍🏼